library/sha256.c

```diff
@@ -54,6 +54,7 @@
 #      if defined(__linux__)
          /* Our preferred method of detection is getauxval() */
 #        include <sys/auxv.h>
+#        include <asm/hwcap.h>
```

Comment:
If there are systems that do not have `asm/hwcap.h`, then this change would break their build.

It looks like [musl](https://packages.ubuntu.com/jammy/arm64/musl-dev/filelist) and [dietlibc](https://packages.ubuntu.com/jammy/arm64/dietlibc-dev/filelist) both ship `sys/auxv.h`, but neither ships `asm/hwcap.h`. (musl ships `bits/hwcap.h`.)

There's an `asm/hwcap.h` in the kernel headers, but that's for compiling kernel code or some rare user land programs that have unusual dependencies on the kernel. In general, kernel headers might not be present, or might not have the correct values for the libc, or might even be only present for the wrong architecture when cross-compiling.

So if CentOS needs an extra header here, I think this needs to be guarded by some `#if`.

Can you point to some documentation about the need for `asm/hwcap.h`? The [man page](https://man7.org/linux/man-pages/man3/getauxval.3.html) only mentions `sys/auxv.h`.

Reply:
>So if CentOS needs an extra header here, I think this needs to be guarded by some `#if`.

Agreed. This was tested on Ubuntu; I don't have a CentOS system set up on an `aarch64` system to test

Reply:
`getauxval()` is indeed defined in `sys/auxv.h`. And `sys/auxv.h` includes `bits/hwcap.h`. `bits/hwcap.h` on CentOS don't contain the any definition of feature flags(`HWCAP_*`). 

:( - I have no idea about other libc implementations(musl or dietlibc).

For the Linux distributions I've encountered, `asm/hwcap.h` must contain the definition of `HWCAP_*`. 

So this is not the CentOS specific case. In the [document's example][0], the need for `asm/hwcap.h` may be explained.


<details>
  <summary>Here is `sys/auxv.h` and `bits/hwcap.h` on CentOS 7</summary>

```
[root@da0acf3938b5 include]# cat -n sys/auxv.h 
     1	/* Access to the auxiliary vector.
     2	   Copyright (C) 2012 Free Software Foundation, Inc.
     3	   This file is part of the GNU C Library.
     4	
     5	   The GNU C Library is free software; you can redistribute it and/or
     6	   modify it under the terms of the GNU Lesser General Public
     7	   License as published by the Free Software Foundation; either
     8	   version 2.1 of the License, or (at your option) any later version.
     9	
    10	   The GNU C Library is distributed in the hope that it will be useful,
    11	   but WITHOUT ANY WARRANTY; without even the implied warranty of
    12	   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    13	   Lesser General Public License for more details.
    14	
    15	   You should have received a copy of the GNU Lesser General Public
    16	   License along with the GNU C Library; if not, see
    17	   <http://www.gnu.org/licenses/>.  */
    18	
    19	#ifndef _SYS_AUXV_H
    20	#define _SYS_AUXV_H 1
    21	
    22	#include <elf.h>
    23	#include <sys/cdefs.h>
    24	#include <bits/hwcap.h>
    25	
    26	__BEGIN_DECLS
    27	
    28	/* Return the value associated with an Elf*_auxv_t type from the auxv list
    29	   passed to the program on startup.  If TYPE was not present in the auxv
    30	   list, returns zero.  */
    31	extern unsigned long int getauxval (unsigned long int __type)
    32	  __THROW __attribute_const__;
    33	
    34	__END_DECLS
    35	
    36	#endif /* sys/auxv.h */
[root@da0acf3938b5 include]# cat -n bits/hwcap.h
     1	/* Defines for bits in AT_HWCAP.
     2	   Copyright (C) 2012 Free Software Foundation, Inc.
     3	   This file is part of the GNU C Library.
     4	
     5	   The GNU C Library is free software; you can redistribute it and/or
     6	   modify it under the terms of the GNU Lesser General Public
     7	   License as published by the Free Software Foundation; either
     8	   version 2.1 of the License, or (at your option) any later version.
     9	
    10	   The GNU C Library is distributed in the hope that it will be useful,
    11	   but WITHOUT ANY WARRANTY; without even the implied warranty of
    12	   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    13	   Lesser General Public License for more details.
    14	
    15	   You should have received a copy of the GNU Lesser General Public
    16	   License along with the GNU C Library; if not, see
    17	   <http://www.gnu.org/licenses/>.  */
    18	
    19	#ifndef _SYS_AUXV_H
    20	# error "Never include <bits/hwcap.h> directly; use <sys/auxv.h> instead."
    21	#endif
    22	
    23	/* No bits defined for this architecture.  */
[root@da0acf3938b5 include]# cat /etc/os-release 
NAME="CentOS Linux"
VERSION="7 (AltArch)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (AltArch)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7:server"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

[root@da0acf3938b5 include]# 
```
</details>

[0]: https://www.kernel.org/doc/html/latest/arm64/cpu-feature-registers.html#appendix-i-example

Reply:
@yuhaoth in your PR you are including both `sys/auxv.h` and `asm/hwcap.h`:
https://github.com/Mbed-TLS/mbedtls/pull/6895/files#diff-5a609345221730119df43d0f62be2c30acc8959c6f9bdb383e4f36a1bab25385R48-R51

What are your thoughts on this?

Reply:
Note that OpenSSL only includes `auxv.h`, not `hwcap.h` - see https://github.com/openssl/openssl/blob/master/crypto/armcap.c

Reply:
OpenSSL defines `HWCAP_*` here:

https://github.com/openssl/openssl/blob/cd870db16348d0d09cb05b7393cf9281509c7795/crypto/armcap.c#L198

Reply:
So it turns out the reason CentOS doesn't define `HWCAP_SHA2` except in kernel headers is simple: its version of Glibc is just too old.

Using kernel headers is not the right solution: it'll break if they're not installed, or when cross-compiling, or in the hypothetical case where the libc uses different constants (unlikely, but such things have happened in the past).

I think the right solution is for us to define the constants manually (like OpenSSL does), though we should use the constants defined by `sys/auxv.h` if they're present.

@valord577 Can you please try https://github.com/Mbed-TLS/mbedtls/pull/7121 ?

Reply:
> @yuhaoth in your PR you are including both `sys/auxv.h` and `asm/hwcap.h`: https://github.com/Mbed-TLS/mbedtls/pull/6895/files#diff-5a609345221730119df43d0f62be2c30acc8959c6f9bdb383e4f36a1bab25385R48-R51
> 
> What are your thoughts on this?

I think we should define `MBEDTLS_HWCAP_*` . I do it in #7078 .

In #6895 , it is out of scope. I just follow exist code and [kernel document](https://github.com/torvalds/linux/blob/master/Documentation/arm64/cpu-feature-registers.rst)